### ISSUE TITLE
`AlchemiscaleClient` `get_tasks_status` and `set_tasks_status` slow for many tasks

### ISSUE THREAD

**Issue by dotsdl** [openforcefield/alchemiscale#148]
When trying to reset many tasks, e.g. 1000, from `error` to `waiting`, it can take quite a bit of time to use `get_tasks_status` and `set_tasks_status` to achieve this. These methods currently loop through all the `Task` `ScopedKey`s they are given one at a time to perform their operations.

We should identify ways to speed this up, including:
- a batched approach, e.g. #147 
- an async/await approach, e.g. #146 
- a combination of approaches, or something else

**Comment by dotsdl** [comment]
@hmacdope can you drop any findings you come across on this question here?

**Comment by hmacdope** [comment]
Google [Firebase](https://firebase.google.com/docs/firestore/manage-data/transactions#python) and [Compute](https://cloud.google.com/compute/docs/api/how-tos/batch) seem to have batching for their endpoints while also supporting `async` modes  

[Mailchimp](https://mailchimp.com/developer/marketing/guides/run-async-requests-batch-endpoint/)  also seems to have an async batched workflow as does Meta's ad [REST API ](https://developers.facebook.com/docs/marketing-api/asyncrequests/). 

Some info on doing batching within Neo4j queries themselves is [here](https://towardsdatascience.com/batching-transactions-in-neo4j-1001d12c9a4a)  and [here](https://medium.com/neo4j/5-tips-tricks-for-fast-batched-updates-of-graph-structures-with-neo4j-and-cypher-73c7f693c8cc)  IMO second one is better. These make heavy use of `UNWIND` .  There is also a general guide to Cypher performance [here](https://aaronyuqi.medium.com/neo4j-performance-tips-for-beginners-62ac92425fa1) which recommends dropping unused identifiers which is something I think we do a little bit. 

Another (IMO great) option is to use `APOC` and its `periodic.iterate` functionality documented  [here](https://neo4j.com/labs/apoc/4.4/overview/apoc.periodic/apoc.periodic.iterate)

```cypher 
CALL apoc.periodic.iterate(
  "UNWIND $nodes AS node
   RETURN node",
  "MATCH (n:Label {id: node.id})
   SET n.property = node.property",
  {batchSize: 100, parallel: true, params: {nodes: $nodeList}}
)
```

**Comment by hmacdope** [comment]
This may require the official drivers rather than `py2neo` but something like:

```python
# Assuming you have a list of unique IDs
id_list = [1, 2, 3, 4]

# Connect to Neo4j
driver = GraphDatabase.driver("bolt://localhost:7687", auth=("neo4j", "password"))

# Execute the Cypher query
with driver.session() as session:
    result = session.run("MATCH (n:Label) WHERE n.id IN $idList RETURN n", idList=id_list)
    nodes = result.data()

# Print the retrieved nodes
for node in nodes:
    print(node)

```

May work according to Copilot X

**Comment by hmacdope** [comment]
I will say that most of the batched APIs have a hardcoded limit on the batchsize.

**Comment by dotsdl** [comment]
Awesome, thanks for these notes @hmacdope! This is great fodder for a decision on how to proceed for these methods.

**Comment by dotsdl** [comment]
Closed by #150.
